pkg/auth/webapp/authflow_routes.go

```diff
@@ -56,7 +60,69 @@ const (
 	AuthflowRouteNoAuthenticator = "/authflow/no_authenticator"
 )
 
+type AuthflowNavigatorEndpointsProvider interface {
+	ErrorEndpointURL(uiImpl config.UIImplementation) *url.URL
+}
+
 type AuthflowNavigator struct {
+	Endpoints   AuthflowNavigatorEndpointsProvider
+	UIConfig    *config.UIConfig
+	ErrorCookie *ErrorCookie
+}
+
+func (n *AuthflowNavigator) NavigateError(r *http.Request, u url.URL, err error) *Result {
+	apierror := apierrors.AsAPIError(err)
+
+	recoverable := func() *Result {
+		cookie, err := n.ErrorCookie.SetRecoverableError(r, apierror)
+		if err != nil {
+			panic(err)
+		}
+
+		result := &Result{
+			RedirectURI:      u.String(),
+			NavigationAction: "replace",
+			Cookies:          []*http.Cookie{cookie},
+		}
+
+		return result
+	}
+
+	nonRecoverable := func() *Result {
+		result := &Result{
+			RedirectURI:      u.String(),
+			NavigationAction: "replace",
+		}
+		err := n.ErrorCookie.SetNonRecoverableError(result, apierror)
+		if err != nil {
+			panic(err)
+		}
+
+		return result
+	}
```

Comment:
Can we keep the cookies related code in authflow_controller, and make `NavigateError` only controls the path?